test/e2e/clusterctl_upgrade.go

```diff
@@ -428,7 +430,30 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 			Name:                   workloadClusterName,
 		}, input.E2EConfig.GetIntervals(specName, "wait-cluster")...)
 
-		expectedMachineCount := *controlPlaneMachineCount + calculateExpectedWorkerCount(ctx, managementClusterProxy.GetClient(), workloadClusterUnstructured, coreCAPIStorageVersion)
+		var infraMachinePool *unstructured.Unstructured
+		skipMachinePoolMachines := false
+
+		By("Determining if MachinePool machines should be skipped")
+		machinePool := getMachinePool(ctx, managementClusterProxy.GetClient(), workloadClusterNamespace, workloadClusterName)
```

Comment:
This part of the test only seems to look at one MachinePool while further down we list MachinePools.

This seems to mean that we only support tests with one sort of MachinePools (with or without Machines) and not mixed. If that is the case. Should we document it?

Reply:
That's a good point. I think I can modify the test to iterate over all MachinePools and support clusters with multiple types of MachinePools.